llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp

```diff
@@ -1014,6 +1016,8 @@ void SystemZAsmPrinter::emitADASection() {
                                 MCSymbolRefExpr::create(Sym, OutContext),
                                 OutContext),
           PointerSize);
+      Sym->setExecutable(GOFF::ESD_EXE_CODE);
```

Comment:
See comment above about setExecutable ... it would be preferable if we didn't have to do that all over the place, but could rely on common code setting the symbol types.

Reply:
So it does look like `GOFF::ESD_EXE_CODE` basically tracks the same information as `MCSA_ELF_TypeFunction` (and ESD_EXE_DATA essentially being analogous to that bit not being set). 

Would you prefer just using that field, or creating a new MCSymbolAttribute? It feels kind of odd to have to use an attribute with ELF in its name for a GOFF-specific thing, but I'd also rather not add a new field to the MCSymbolAttribute if only one target will use it?

Reply:
I think it would be OK to use MCSA_ELF_Type.   The Wasm format also uses those, even though it is not exactly ELF ...

Reply:
Sorry, can we take a step back? 

Is if the problem is us setting the Executable flag all over the place, or if the problem is that it seems somewhat similar to `MCSA_ELF_TypeFunction` and `MCSA_ELF_TypeObject`. 

If it's the former, I can see how it is problematic. I did a quick test and it seems that most places actually set it redundantly, so that definitely needs to be cleaned up. My first thought is to override `doFinalization` in the SystemZAsmPrinter, and (for zOS targets only), set the appropriate flag as needed there (and nowhere else).

If it's the latter, then we can piggyback off of that flag, but I don't think we can exactly rely on common code, at least without modifying it somewhat. That flag gets set in the following places (by common code): `emitFunctionHeader`, `emitFunctionEntryLabel`, and `emitGlobalAlias`.

The problem is that we also need to set that flag for functions that are declared but not defined in the current compilation unit, and none of those functions mentioned above are called for function declarations, so we wouldn't be able to entirely piggyback off common code anyways, and if we attempted to do so we would still need to set it in multiple places.

Reply:
I guess in the end the interface to the GOFF writer should closely follow the HLASM language.  Eventually, we want to able to create both a valid GOFF object file and a valid HLASM assembler output file from the same AsmWriter calls.

So the question becomes, how would setting the Executable flag for external symbols be implemented in HLASM source?  Would this be inferred from the symbol's use inside a VCon or RCon expression?  Or would we have to use an explicit HLASM directive like XATTR REF(CODE) ?  This should guide the design of the interface.

If an explicit directive would be needed, this should be implemented via a corresponding AsmWriter call (just like the MCSA_ELF_Type* attributes correspond to the .type directive in ELF assemblers).   In any case, I'd prefer the decision to be made close to the place of usage, rather than in a completely separate pass.

Reply:
It would be the latter (ie. an explicit directive).

The executable flag would look something like `SYMBOLNAME XATTR REFERENCE(CODE)`

The data flag would be `SYMBOLNAME XATTR REFERENCE(DATA)`

Reply:
Thinking about this a bit more, I actually like the current approach of setting this flag in doFinalization once for every external symbol.  (I think you should probably exclude internal symbols there, however.)

Given that this behavior is actually quite a bit different from the ELF .type handling (which applies only to defined symbols), however, I now think it would indeed be better to define some new MCSA_GOFF_ attributes rather than reusing the ELF_Type attributes.  We'll need other GOFF-specific symbol attributes anyway.